analyzers/tests/SonarAnalyzer.UnitTest/Rules/RedundantCastTest.cs

```diff
@@ -21,45 +21,99 @@
 #endif
 
-        [TestMethod]
-        public void RedundantCast_CodeFix() =>
-            builder.AddPaths("RedundantCast.cs").WithCodeFix<RedundantCastCodeFix>().WithCodeFixedPaths("RedundantCast.Fixed.cs").VerifyCodeFix();
+    [TestMethod]
+    public void RedundantCast_CodeFix() =>
+        builder.AddPaths("RedundantCast.cs").WithCodeFix<RedundantCastCodeFix>().WithCodeFixedPaths("RedundantCast.Fixed.cs").VerifyCodeFix();
 
-        [TestMethod]
-        public void RedundantCast_DefaultLiteral() =>
-            builder.AddSnippet(@"
-using System;
-public static class MyClass
-{
-    public static void RunAction(Action action)
+    [TestMethod]
+    public void RedundantCast_DefaultLiteral() =>
+        builder.AddSnippet("""
+            using System;
+            public static class MyClass
+            {
+                public static void RunAction(Action action)
+                {
+                    bool myBool = (bool)default; // FN - the cast is unneeded
+                    RunFunc(() => { action(); return default; }, (bool)default); // should not raise because of the generic the cast is mandatory
+                    RunFunc<bool>(() => { action(); return default; }, (bool)default); // FN - the cast is unneeded
+                }
+
+                 public static T RunFunc<T>(Func<T> func, T returnValue = default) => returnValue;
+            }
+            """).WithLanguageVersion(LanguageVersion.CSharp7_1).Verify();
+
+    private static IEnumerable<object[]> NullableTestData => new[]
     {
-        bool myBool = (bool)default; // FN - the cast is unneeded
-        RunFunc(() => { action(); return default; }, (bool)default); // should not raise because of the generic the cast is mandatory
-        RunFunc<bool>(() => { action(); return default; }, (bool)default); // FN - the cast is unneeded
-    }
+        new object[] { """_ = (string)"Test";""", false, false },
+        new object[] { """_ = (string?)"Test";""", true, false },
+        new object[] { """_ = (string)null;""", true, true },
+        new object[] { """_ = (string?)null;""", true, true },
+        new object[] { """_ = (string)nullable;""", true, false },
+        new object[] { """_ = (string?)nullable;""", false, false },
+        new object[] { """_ = (string)nonNullable;""", false, false },
+        new object[] { """_ = (string?)nonNullable;""", true, false },
+        new object[] { """if (nullable != null) _ = (string)nullable;""", false, false },
+        new object[] { """if (nullable != null) _ = (string?)nullable;""", true, false },
+        new object[] { """if (nonNullable == null) _ = (string)nonNullable;""", true, false },
+        new object[] { """if (nonNullable == null) _ = (string?)nonNullable;""", false, false },
+    };
+
+    private static IEnumerable<object[]> NullableTestDataWithFlowState =>
+        NullableTestData.Select(x => new[] { x[0], x[1] });
```

Comment:
If you always do the select, you can also strong type the return type of `NullableTestData`:
```cs
private static IEnumerable<(string snippet, bool compliantWithFlowState, bool compliantWithoutFlowState)> NullableTestData()
```